# CPSC-483 Project 3: Classification
## FA 2021 - Section 02
### Armanul Ambia, Iftekharul Islam

### __Experiment 1__ : Examine the dataset from bank-additional-full.csv.

In [1]:
import pandas as pd

df = pd.read_csv('bank-additional-full.csv', sep = ';')
df.head()

age        job  marital    education  default housing loan    contact  \
0   56  housemaid  married     basic.4y       no      no   no  telephone   
1   57   services  married  high.school  unknown      no   no  telephone   
2   37   services  married  high.school       no     yes   no  telephone   
3   40     admin.  married     basic.6y       no      no   no  telephone   
4   56   services  married  high.school       no      no  yes  telephone   

  month day_of_week  ...  campaign  pdays  previous     poutcome emp.var.rate  \
0   may         mon  ...         1    999         0  nonexistent          1.1   
1   may         mon  ...         1    999         0  nonexistent          1.1   
2   may         mon  ...         1    999         0  nonexistent          1.1   
3   may         mon  ...         1    999         0  nonexistent          1.1   
4   may         mon  ...         1    999         0  nonexistent          1.1   

   cons.price.idx  cons.conf.idx  euribor3m  nr.employed   y  
0          93.994          -36.4      4.857       5191.0  no  
1          93.994          -36.4      4.857       5191.0  no  
2          93.994          -36.4      4.857       5191.0  no  
3          93.994          -36.4      4.857       5191.0  no  
4          93.994          -36.4      4.857       5191.0  no  

[5 rows x 21 columns]

The dataset from bank-additional-full.csv is read using the Pandas library. The first few rows of the dataframe are displayed above and show that there are 20 variables used to determine whether a client will subscribe to a term deposit. The final column consists of variable 'y' to indicate whether the individual represented by that row subscribed. This is the value we will be classifying.

### __Experiment 2__ : Split the features and target values into separate training and test sets.

In [2]:
from sklearn.model_selection import train_test_split

training, test = train_test_split(df, train_size = 0.9, test_size = 0.1, random_state=(2021-10-25))

print(f'Size of training set: {len(training)}')
training.head()

Size of training set: 37069


age         job   marital            education default  housing  \
22975   56     retired   married  professional.course      no      yes   
14746   26  technician    single  professional.course      no  unknown   
12505   38      admin.  divorced          high.school      no      yes   
6801    33      admin.  divorced    university.degree      no       no   
18389   29      admin.   married          high.school      no       no   

          loan    contact month day_of_week  ...  campaign  pdays  previous  \
22975       no   cellular   aug         mon  ...         5    999         0   
14746  unknown   cellular   jul         wed  ...         1    999         0   
12505       no   cellular   jul         mon  ...         1    999         0   
6801        no  telephone   may         wed  ...         2    999         0   
18389       no   cellular   jul         thu  ...         3    999         0   

          poutcome emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  \
22975  nonexistent          1.4          93.444          -36.1      4.965   
14746  nonexistent          1.4          93.918          -42.7      4.957   
12505  nonexistent          1.4          93.918          -42.7      4.960   
6801   nonexistent          1.1          93.994          -36.4      4.857   
18389  nonexistent          1.4          93.918          -42.7      4.968   

       nr.employed    y  
22975       5228.1   no  
14746       5228.1  yes  
12505       5228.1   no  
6801        5191.0   no  
18389       5228.1   no  

[5 rows x 21 columns]

In [3]:
print(f'Size of test set: {len(test)}')
test.head()

Size of test set: 4119


age          job   marital          education  default housing loan  \
37215   38       admin.    single  university.degree       no     yes   no   
576     41       admin.  divorced        high.school       no      no   no   
14186   35  blue-collar   married        high.school  unknown     yes  yes   
4527    26  blue-collar   married           basic.4y       no     yes  yes   
3900    43       admin.  divorced        high.school       no      no   no   

         contact month day_of_week  ...  campaign  pdays  previous  \
37215   cellular   aug         fri  ...         1    999         0   
576    telephone   may         tue  ...         4    999         0   
14186  telephone   jul         mon  ...         3    999         0   
4527   telephone   may         tue  ...         2    999         0   
3900   telephone   may         mon  ...         2    999         0   

          poutcome emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  \
37215  nonexistent         -2.9          92.201          -31.4      0.881   
576    nonexistent          1.1          93.994          -36.4      4.857   
14186  nonexistent          1.4          93.918          -42.7      4.962   
4527   nonexistent          1.1          93.994          -36.4      4.856   
3900   nonexistent          1.1          93.994          -36.4      4.858   

       nr.employed   y  
37215       5076.2  no  
576         5191.0  no  
14186       5228.1  no  
4527        5191.0  no  
3900        5191.0  no  

[5 rows x 21 columns]

The dataset has been separated into a training set and a test set using the scikit-learn library. The training set consists of 90% of the data while the test set is made up of the remaining 10%. A random state is passed in as well to ensure the data is shuffled the same way every time the program runs to ensure reproducibility. To confirm the dataset split properly, the first few rows of both sets as well as the size of each set are displayed above.

### __Experiment 3__ : Process the training and test sets before they are used.

### Experiment 3a : Discard the duration column.

In [4]:
training = training.drop(columns = 'duration')
test = test.drop(columns = 'duration')

print("Columns in training set:")
print(training.columns)

print("\nColumns in test set:")
print(test.columns)

Columns in training set:
Index(['age', 'job', 'marital', 'education', 'default', 'housing', 'loan',
       'contact', 'month', 'day_of_week', 'campaign', 'pdays', 'previous',
       'poutcome', 'emp.var.rate', 'cons.price.idx', 'cons.conf.idx',
       'euribor3m', 'nr.employed', 'y'],
      dtype='object')

Columns in test set:
Index(['age', 'job', 'marital', 'education', 'default', 'housing', 'loan',
       'contact', 'month', 'day_of_week', 'campaign', 'pdays', 'previous',
       'poutcome', 'emp.var.rate', 'cons.price.idx', 'cons.conf.idx',
       'euribor3m', 'nr.employed', 'y'],
      dtype='object')


The duration of a call drastically impacts the results of the model. For example, if duration = 0, y becomes "no". Because of this, it is discarded to obtain an accurate model. A list of column names for both the training and test sets are shown above to ensure duration does not exist among them.

### Experiment 3b : Set aside feature y for use in training and testing since it is the target response. Then, drop it from the features.

In [5]:
# Storing y column
training_y = training['y']
test_y = test['y']

# Dropping y column from both training and test sets
training = training.drop(columns = 'y')
test = test.drop(columns = 'y')

print("Columns in training set:")
print(training.columns)

print("\nColumns in test set:")
print(test.columns)

Columns in training set:
Index(['age', 'job', 'marital', 'education', 'default', 'housing', 'loan',
       'contact', 'month', 'day_of_week', 'campaign', 'pdays', 'previous',
       'poutcome', 'emp.var.rate', 'cons.price.idx', 'cons.conf.idx',
       'euribor3m', 'nr.employed'],
      dtype='object')

Columns in test set:
Index(['age', 'job', 'marital', 'education', 'default', 'housing', 'loan',
       'contact', 'month', 'day_of_week', 'campaign', 'pdays', 'previous',
       'poutcome', 'emp.var.rate', 'cons.price.idx', 'cons.conf.idx',
       'euribor3m', 'nr.employed'],
      dtype='object')


Column y is the target response we want to find using our model. Because of this, we want to keep it separate from our features. The column names of the training and test sets have been printed once more to show the y column has properly dropped.

### __Experiment 4 :__ Take input variables from bank-additional-names.txt to determine whether a client will subscribe to a term deposit.

In [6]:
# Create subset of training set and test sets using only bank client data
training_inputs = training[['job', 'marital', 'age', 'education', 'default', 'housing', 'loan']]
test_inputs = test[['job', 'marital', 'age', 'education', 'default', 'housing', 'loan']]

# Convert categorical values into dummy variables so that entire dataframe is numeric
training_inputs = pd.get_dummies(training_inputs, columns = ['job', 'marital', 'education', 'default', 'housing', 'loan'], 
                          prefix_sep='.', drop_first = True)
test_inputs = pd.get_dummies(test_inputs, columns = ['job', 'marital', 'education', 'default', 'housing', 'loan'], 
                          prefix_sep='.', drop_first = True)


The features that will be used for the model have been identified under "bank client data" in the bank-additional-names.txt file. Out of the seven features, six are categorical. To make them treatable as vectors, we turn them into dummy variables using the Pandas get_dummies() function. We also pass drop_first as True to get k-1 dummies out of the k categorical levels in order to ensure we have no issues with collinear features.

### __Experiment 5 :__ Fit a Categorical Naive Bayes classifier to the features identified in Experiment 4. Then, score it on both the training and test sets to test its accuracy.

In [7]:
import numpy as np
from sklearn.naive_bayes import CategoricalNB
clf = CategoricalNB()

clf.fit(training_inputs, training_y)

print(f"Training Score: {clf.score(training_inputs, training_y)}")
print(f"Test Score: {clf.score(test_inputs, test_y)}")

Training Score: 0.8818689470986538
Test Score: 0.8788540907987376


The Categorical Naive Bayes classifier seems to be very accurate. The training and test scores were both approximately .88 indicating a high correlation between the predicted and actual outcomes. The model was fit using the training set and training y values and then tested on both the training and test sets. It is interesting, however, that the values are so close even though the test set had not been introduced to the model and the training set was used to fit the model. 

### __Experiment 6__: Consider how many categories are created if age is assumed to be categorical instead of a quantiative predictor.

In [8]:
clf.category_count_[0].shape

(2, 99)

By examining the CategoricalNB model, we can see that for age, there are a total of 99 categories created. Since this is a quantitative predictor, this is not reasonable and the number of categories should be limited. A category for each age group is not needed.

### __Experiment 7__: Split ages into bins, one per decade. Verify the number of bins, and retrain your classifier model.

In [9]:
training_inputs_age = training_inputs
test_inputs_age = test_inputs

training_inputs_age['age'] = training_inputs_age['age'] // 10 
test_inputs_age['age'] = test_inputs_age['age'] // 10 

clf_age = CategoricalNB()
clf_age.fit(training_inputs_age, training_y)

# Verify that there are 10 bins (1-9) for age
print(f"Training Age Minimum Value: {min(training_inputs_age['age'])}")
print(f"Training Age Maximum Value: {max(training_inputs_age['age'])}")
print(f"Number of bins: {len(clf_age.category_count_[0][0])}")

print(f"\nNew Training Score: {clf_age.score(training_inputs_age, training_y)}")
print(f"New Test Score: {clf_age.score(test_inputs_age, test_y)}")

Training Age Minimum Value: 1
Training Age Maximum Value: 9
Number of bins: 10

New Training Score: 0.8812484825595511
New Test Score: 0.8786113134255887


The ages have been split into bins representing a single decade. This is accomplished using the integer division operator so that the age column now has 10 bins ranging from 1 to 9. Our new scores, however, are still very similar to the previous classification. The values are slightly higher but no noticeable difference. Again, the test score is almost identical to the training score.

### __Experiment 8__: Repeat Experiment 5 with KNN classifier.

In [10]:
from sklearn.neighbors import KNeighborsClassifier as knn

neigh = knn(n_neighbors = 5)
neigh.fit(training_inputs, training_y)

print(f"Training Score: {neigh.score(training_inputs, training_y)}")
print(f"Test Score: {neigh.score(test_inputs, test_y)}")


Training Score: 0.8738298848094095
Test Score: 0.868657441126487


We used a KNN classifier with 5 neighbors in order to fit the training set with the training outcomes. After scoring both the trianing and test set, we have obtained nearly identical values to the previous 2 experiments. This indicates that the models are not giving correct outputs or the data needs to be adjusted before using these models.

### __Experiment 9__: Check the test set for the number of values with response 0 and response 1.

In [11]:
yes = test_y.value_counts().yes
no = test_y.value_counts().no

print(f"No responses in test set: {no}")
print(f"Yes responses in test set: {yes}")


total_length = len(test_y)
print(f"\n\nTotal responses in test set: {total_length}")
# simulate accuracy if all were predicted as no
print(f"Simulated Accuracy if all responses are no: {no/total_length}")

No responses in test set: 3646
Yes responses in test set: 473


Total responses in test set: 4119
Simulated Accuracy if all responses are no: 0.8851663025006069


By examining the number of yes (1) responses and no (0) responses in the test set, we found that there are far fewer yes responses than no responses. If we simply take the number of outcomes and assume all are 0 with our prediction. The accuracy of this simulation would be .885 and nearly identical to the previous classification model scores. This indicates that the data must be adjusted before being able to accurately run these models. 

### __Experiment 10__: Create the dumb classifier vector and then create confusion matrix to find AUC.

In [12]:
import numpy as np
from sklearn.metrics import confusion_matrix as cm

# Convert yes and no to 1 and 0 respectively
temp_y = test_y.eq('yes').mul(1)

# Create 0 vector
dummy_y = np.zeros_like(temp_y)

# Create confusion matrix with dummy_y and temp_y
matrix = cm(temp_y.values, dummy_y)

# Find AUC score
from sklearn.metrics import roc_auc_score

print(f"Confusion Matrix:\n {matrix}")
print(f"AUC: {roc_auc_score(temp_y, dummy_y)}")

Confusion Matrix:
 [[3646    0]
 [ 473    0]]
AUC: 0.5


In this experiment we created a dummy vector of zeros equal in length to the test set. We used this to create a confusion matrix and receive the AUC score. The confusion matrix follows the following format:
\begin{bmatrix}
TN & FP \\
FN & TP
\end{bmatrix}
<br>
<center>
    <strong>TP</strong> = True Negative<br>
    <strong>FN</strong> = False Negative<br>
    <strong>FP</strong> = False Positive<br>
    <strong>TP</strong> = True Positive<br>
</center>

This means that our matrix produces 3646 True Negatives. This is accurate since there were 3646 no's found in previous experiment. We also had 473 false negatives. This means that we predicted no (since all the predictions were no) while there were 473 yes's. This is also accurate according to the previous experiment's yes count.

Using the dummy vector and the test outcome values we can compute the Area Under the Curve (AUC) to be 0.5. This is expected since we used a constant prediction value; an AUC of 0.5 indicates that the model really does not predict a value or simply returns a constant. Graphically, this would mean the curve is simple the line y = x indicating a constant slope with half the area of the graph underneath.

### __Experiment 11__: Create confusion matrices and compute the AUC for each of the classifiers in Experiments 7 and 8.

In [13]:
# Convert yes and no to 1 and 0, respectively
temp_y = test_y.eq('yes').mul(1)

# Create predictions for exp 7 using categorical naive bayes classifier
pred_y = clf_age.predict(test_inputs_age)
pred_y[pred_y=='yes'] = 1
pred_y[pred_y=='no'] = 0
pred_y = pred_y.astype(int)

# Create confusion matrix for exp 7 with pred_y and temp_y
matrix = cm(temp_y.values, pred_y)

print(f"Exp. 7 - Confusion Matrix:\n {matrix}")
print(f"Exp. 7 - AUC: {roc_auc_score(temp_y.values, pred_y)}")

# Create predictions for exp 8 using knn nearest neighbors classifier
pred_y = neigh.predict(test_inputs)
pred_y[pred_y=='yes'] = 1
pred_y[pred_y=='no'] = 0
pred_y = pred_y.astype(int)

# Create confusion matrix for exp 8 with pred_y and temp_y
matrix = cm(temp_y.values, pred_y)
print(f"\nExp. 8 - Confusion Matrix:\n {matrix}")
print(f"Exp. 8 - AUC: {roc_auc_score(temp_y, pred_y)}")

Exp. 7 - Confusion Matrix:
 [[3559   87]
 [ 413   60]]
Exp. 7 - AUC: 0.5514940639862504

Exp. 8 - Confusion Matrix:
 [[3537  109]
 [ 432   41]]
Exp. 8 - AUC: 0.5283924924531387


After retraining the data we are able to see the confusion matrix and the AUC score for the CategoricalNB classifier and the KNN classifier. While the KNN classifier performed slightly better than the CategoricalNB, both models still had very low AUC scores close to .5 indicating that the prediction model is still almost constant and not giving an accurate result.

### __Experiment 12__: Oversample to add weight to yes values.

In [14]:
# Oversample training set
no_count, yes_count = training_y.value_counts()
training_yes = training_inputs[training_y == 'yes']
training_no = training_inputs[training_y == 'no']
training_yes_y = training_y[training_y == 'yes']
training_no_y = training_y[training_y == 'no']

df_yes = training_yes.sample(no_count, replace=True, random_state=(2021-10-25))
df_yes_y = training_yes_y.sample(no_count, replace=True, random_state=(2021-10-25))

oversampled_training = pd.concat([training_no, df_yes], axis=0)
oversampled_training_y = pd.concat([training_no_y, df_yes_y], axis=0)

# Oversample test set
no_count, yes_count = test_y.value_counts()
test_yes = test_inputs[test_y == 'yes']
test_no = test_inputs[test_y == 'no']
test_yes_y = test_y[test_y == 'yes']
test_no_y = test_y[test_y == 'no']

df_yes = test_yes.sample(no_count, replace=True, random_state=(2021-10-25))
df_yes_y = test_yes_y.sample(no_count, replace=True, random_state=(2021-10-25))

oversampled_test = pd.concat([test_no, df_yes], axis=0)
oversampled_test_y = pd.concat([test_no_y, df_yes_y], axis=0)

oversampled_training[oversampled_training_y == 'yes'].head()

age  job.blue-collar  job.entrepreneur  job.housemaid  job.management  \
39026    6                0                 0              1               0   
36770    5                1                 0              0               0   
14050    4                0                 0              0               0   
30437    5                0                 0              1               0   
4609     4                0                 0              0               0   

       job.retired  job.self-employed  job.services  job.student  \
39026            0                  0             0            0   
36770            0                  0             0            0   
14050            0                  1             0            0   
30437            0                  0             0            0   
4609             0                  0             0            0   

       job.technician  ...  education.illiterate  \
39026               0  ...                     0   
36770               0  ...                     0   
14050               0  ...                     0   
30437               0  ...                     0   
4609                0  ...                     0   

       education.professional.course  education.university.degree  \
39026                              0                            0   
36770                              0                            0   
14050                              1                            0   
30437                              1                            0   
4609                               0                            0   

       education.unknown  default.unknown  default.yes  housing.unknown  \
39026                  0                0            0                0   
36770                  0                0            0                0   
14050                  0                0            0                0   
30437                  0                0            0                0   
4609                   0                0            0                0   

       housing.yes  loan.unknown  loan.yes  
39026            0             0         0  
36770            1             0         0  
14050            0             0         0  
30437            0             0         0  
4609             0             0         0  

[5 rows x 28 columns]

In [15]:
oversampled_training[oversampled_training_y == 'no'].head()

age  job.blue-collar  job.entrepreneur  job.housemaid  job.management  \
22975    5                0                 0              0               0   
12505    3                0                 0              0               0   
6801     3                0                 0              0               0   
18389    2                0                 0              0               0   
8091     3                0                 0              0               0   

       job.retired  job.self-employed  job.services  job.student  \
22975            1                  0             0            0   
12505            0                  0             0            0   
6801             0                  0             0            0   
18389            0                  0             0            0   
8091             0                  0             0            0   

       job.technician  ...  education.illiterate  \
22975               0  ...                     0   
12505               0  ...                     0   
6801                0  ...                     0   
18389               0  ...                     0   
8091                0  ...                     0   

       education.professional.course  education.university.degree  \
22975                              1                            0   
12505                              0                            0   
6801                               0                            1   
18389                              0                            0   
8091                               0                            1   

       education.unknown  default.unknown  default.yes  housing.unknown  \
22975                  0                0            0                0   
12505                  0                0            0                0   
6801                   0                0            0                0   
18389                  0                0            0                0   
8091                   0                0            0                0   

       housing.yes  loan.unknown  loan.yes  
22975            1             0         0  
12505            1             0         0  
6801             0             0         0  
18389            0             0         0  
8091             1             0         0  

[5 rows x 28 columns]

In [16]:
print(f"Number of Yes Values in new training set: {oversampled_training[oversampled_training_y == 'yes'].shape[0]}" )
print(f"Number of No Values in new training set: {oversampled_training[oversampled_training_y == 'no'].shape[0]}" )

Number of Yes Values in new training set: 32902
Number of No Values in new training set: 32902


In this experiment we oversampled the data by adding more yes values. We know that the yes count is far lower than the no count and we duplicated yes outcomes to grow the training and test sets until there are an equal number of yes and no values. The resulting training set has 32902 no values (unchanged) but now also has 32902 yes values. The mapping is created using the training outcomes vector. In addition, we scaled the test set, training outcomes, and test outcomes so that all four sets have the correct dimensions for further analysis.

### __Experiment 13__: Retrain both classifiers and find the score, confusion matrix, and AUC.

In [17]:
# Categorical Naive Bayes
oversampled_training_age = oversampled_training
oversampled_test_age = oversampled_test
oversampled_training_age['age'] = oversampled_training_age['age'] // 10
oversampled_test_age['age'] = oversampled_test_age['age'] // 10

clf_age = CategoricalNB()
clf_age.fit(oversampled_training_age, oversampled_training_y)

print(f"CategoricalNB - Training Score: {clf_age.score(oversampled_training_age, oversampled_training_y)}")
print(f"CategoricalNB - Test Score: {clf_age.score(oversampled_test_age, oversampled_test_y)}")

temp_y = oversampled_test_y.eq('yes').mul(1)
pred_y = clf_age.predict(oversampled_test_age)
pred_y[pred_y=='yes'] = 1
pred_y[pred_y=='no'] = 0
pred_y = pred_y.astype(int)
matrix = cm(temp_y.values, pred_y)

print(f"CategoricalNB - Confusion Matrix:\n {matrix}")
print(f"CategoricalNB - AUC: {roc_auc_score(temp_y, pred_y)}")

# K-Nearest Neighbors
neigh = knn(n_neighbors = 5)
neigh.fit(oversampled_training, oversampled_training_y)

print(f"\n\nKNN - Training Score: {neigh.score(oversampled_training, oversampled_training_y)}")
print(f"KNN - Test Score: {neigh.score(oversampled_test, oversampled_test_y)}")

pred_y = neigh.predict(oversampled_test)
pred_y[pred_y=='yes'] = 1
pred_y[pred_y=='no'] = 0
pred_y = pred_y.astype(int)

# create confusion matrix for exp 8 with pred_y and temp_y
matrix = cm(temp_y.values, pred_y)
print(f"KNN - Confusion Matrix:\n {matrix}")
print(f"KNN - AUC: {roc_auc_score(temp_y, pred_y)}")

CategoricalNB - Training Score: 0.5997203817397119
CategoricalNB - Test Score: 0.5877674163466813
CategoricalNB - Confusion Matrix:
 [[2088 1558]
 [1448 2198]]
CategoricalNB - AUC: 0.5877674163466813


KNN - Training Score: 0.6136101148866331
KNN - Test Score: 0.5792649478880966
KNN - Confusion Matrix:
 [[2108 1538]
 [1530 2116]]
KNN - AUC: 0.5792649478880965


After oversampling the data we can see an improvement in both the CategoricalNB classifier as well as the KNN classifier. The AUC score for the CategoricalNB classifier jumped up to .587 and the KNN AUC also jumped up to .579. In addition, the confusion matrix shows many more values for our true positives and true negatives. There are still a significant number of false positives and false negatives, however. Since these values are relatively low even after oversampling the data, the bank client data may not be a good set of features to predict the outcomes. Possible solutions would be to choose different features or select different models (or both).

### __Experiment 14__: Retrain with social and economic context attributes and find the score, confusion matrix, and AUC.

In [18]:
training_inputs = training[['emp.var.rate','cons.price.idx','cons.conf.idx','euribor3m','nr.employed']]
test_inputs = test[['emp.var.rate','cons.price.idx','cons.conf.idx','euribor3m','nr.employed']]

from sklearn.naive_bayes import GaussianNB
glf = GaussianNB()
glf.fit(training_inputs, training_y)

print(f"Gaussian - Training Score: {glf.score(training_inputs, training_y)}")
print(f"Gaussian - Test Score: {glf.score(test_inputs, test_y)}")

temp_y = test_y.eq('yes').mul(1)
pred_y = glf.predict(test_inputs)
pred_y[pred_y=='yes'] = 1
pred_y[pred_y=='no'] = 0
pred_y = pred_y.astype(int)
matrix = cm(temp_y.values, pred_y)

print(f"Gaussian - Confusion Matrix:\n {matrix}")
print(f"Gaussian - AUC: {roc_auc_score(temp_y.values, pred_y)}")

Gaussian - Training Score: 0.7199816558310178
Gaussian - Test Score: 0.7193493566399611
Gaussian - Confusion Matrix:
 [[2634 1012]
 [ 144  329]]
Gaussian - AUC: 0.7089978997517046


In this experiment, we used different features corresponding to social and economic context attributes. Using this new data set, we were able to fit a Gaussian classifier since there are no categorical features. Using this new model and a new training and test set, we were able to get much higher scores than the previous models. Our Gaussian score was .719, which was lower than the scores found previously. However, the AUC was much higher at .709 indicating a much better classification model and more accurate predictions. 

### __Experiment 15__: Retrain with oversampled data using social and economic context attributes and find the score, confusion matrix, and AUC.

In [19]:
# Oversample training set
no_count, yes_count = training_y.value_counts()
training_yes = training_inputs[training_y == 'yes']
training_no = training_inputs[training_y == 'no']
training_yes_y = training_y[training_y == 'yes']
training_no_y = training_y[training_y == 'no']

df_yes = training_yes.sample(no_count, replace=True, random_state=(2021-10-25))
df_yes_y = training_yes_y.sample(no_count, replace=True, random_state=(2021-10-25))

oversampled_training = pd.concat([training_no, df_yes], axis=0)
oversampled_training_y = pd.concat([training_no_y, df_yes_y], axis=0)

# Oversample test set
no_count, yes_count = test_y.value_counts()
test_yes = test_inputs[test_y == 'yes']
test_no = test_inputs[test_y == 'no']
test_yes_y = test_y[test_y == 'yes']
test_no_y = test_y[test_y == 'no']

df_yes = test_yes.sample(no_count, replace=True, random_state=(2021-10-25))
df_yes_y = test_yes_y.sample(no_count, replace=True, random_state=(2021-10-25))

oversampled_test = pd.concat([test_no, df_yes], axis=0)
oversampled_test_y = pd.concat([test_no_y, df_yes_y], axis=0)

# Retrain
glf = GaussianNB()
glf.fit(oversampled_training, oversampled_training_y)

print(f"Gaussian - Training Score: {glf.score(oversampled_training, oversampled_training_y)}")
print(f"Gaussian - Test Score: {glf.score(oversampled_test, oversampled_test_y)}")

temp_y = oversampled_test_y.eq('yes').mul(1)
pred_y = glf.predict(oversampled_test)
pred_y[pred_y=='yes'] = 1
pred_y[pred_y=='no'] = 0
pred_y = pred_y.astype(int)
matrix = cm(temp_y.values, pred_y)

print(f"Gaussian - Confusion Matrix:\n {matrix}")
print(f"Gaussian - AUC: {roc_auc_score(temp_y, pred_y)}")

Gaussian - Training Score: 0.7178287034222843
Gaussian - Test Score: 0.7111903455842019
Gaussian - Confusion Matrix:
 [[2634 1012]
 [1094 2552]]
Gaussian - AUC: 0.7111903455842019


In this experiment, we oversampled the data following a similar method to Experiment 12. This should theoretically produce better results since the data goes from containing many more no values than yes values to the same number of yes and no values. However, the model gives only a slight improvement in the score as well as the AUC. Still, the new social and economic set of features with a Gaussian classifier produce a much better model than the bank and client data features with KNN and CategoricalNB classifiers.